### PR TITLE
Expose the ioapic and pic features API

### LINKED ISSUE #8066: Expose the ioapic and pic features API
**Is your feature request related to a problem? Please describe**:
When trying to enable TDX support in Kubevirt, we found that ioapic and pic features can't
be modified in the VMI yaml file which mean the ioapic was fixed to 'kvm' and pic was always 'on'. 
But for the TDX support, the ioapic only support 'qemu' (split mode) and pic should be 'off'.

**Describe the solution you'd like**:

We need the support to configure the ioapic and pic feature in VM/VMI yaml file just as below:

> features:
      pic:
        enabled: false
      ioapic:
        driver: qemu

**Describe alternatives you've considered**:

**Additional context**:
(https://www.intel.com/content/www/us/en/developer/articles/technical/intel-trust-domain-extensions.html)

### PR DESCRIPTION
**Pull request by xiaobo55x**
**What this PR does / why we need it**:
When trying to enable TDX support in Kubevirt, we found that ioapic and pic features can't
be modified in the VMI yaml file which mean the ioapic was fixed to 'kvm' and pic was always 'on'.
But for the TDX support, the ioapic only support 'qemu' (split mode) and pic should be 'off'.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8066 

**Special notes for your reviewer**:

**Release note**:

```release-note
New feature APIs are enabled for ioapic and pic when define a VM or VMI.
features:
   pic:
      enabled: false
   ioapic:
      driver: qemu
```

### CODE REVIEW

File: api/openapi-spec/swagger.json [hunk 1]
```diff
@@ -15837,6 +15837,15 @@
      }
     }
    },
+   "v1.FeatureIOAPIC": {
+    "type": "object",
+    "properties": {
+     "driver": {
+      "description": "Configure the guest I/O APIO mode('kvm' or 'qemu').",
```
Comment (optimistyzy) [inline api/openapi-spec/swagger.json:15844]: Besides QEMU, we also support Rust based VMM, e.g., Cloud hypervisor

File: pkg/virt-launcher/virtwrap/converter/converter_test.go [hunk 2]
```diff
@@ -3234,6 +3234,84 @@ var _ = Describe("Converter", func() {
...
+
+		It("should not configure the ioapic feature", func() {
+			domain := vmiToDomain(vmi, c)
+			Expect(domain.Spec.Features.IOAPIC).To(BeNil())
+		})
+
+		It("should configure the ioapic feature", func() {
+			vmi.Spec.Domain.Features = &v1.Features{
+				IOAPIC: &v1.FeatureIOAPIC{
+					Driver: "qemu",
```
Comment (optimistyzy) [inline pkg/virt-launcher/virtwrap/converter/converter_test.go:3268]: Same question, will we only limit to QEMU?

File: pkg/virt-operator/resource/generate/components/validations_generated.go [hunk 3]
```diff
@@ -19094,6 +19199,14 @@ var CRDsValidation map[string]string = map[string]string{
                       type: boolean
                   type: object
               type: object
+            preferredIoapic:
+              description: PreferredIoapic optionally enables and configures IOAPIC
+                feature
+              properties:
+                driver:
+                  description: Configure the guest I/O APIO mode('kvm' or 'qemu').
```
Comment (optimistyzy) [inline pkg/virt-operator/resource/generate/components/validations_generated.go:19207]: APIO ->APIC ?

File: staging/src/kubevirt.io/api/core/v1/schema_swagger_generated.go [hunk 4]
```diff
@@ -592,6 +594,12 @@ func (FeatureKVM) SwaggerDoc() map[string]string {
 	}
 }
 
+func (FeatureIOAPIC) SwaggerDoc() map[string]string {
+	return map[string]string{
+		"driver": "Configure the guest I/O APIO mode('kvm' or 'qemu').",
```
Comment (optimistyzy) [inline staging/src/kubevirt.io/api/core/v1/schema_swagger_generated.go:599]: APIO -> APIC?

**Comment by alicefr** [comment]
@xiaobo55x do you see other reasons for exposing those values to KubeVirt API? Otherwise, those values could be set when txd is requested.

File: api/openapi-spec/swagger.json (hunk 1 above)
Comment (xiaobo55x) [reply api/openapi-spec/swagger.json:15844]: Currently, the patch only consider using libvirt to start the VM.  More works were needed to cover the Cloud hypervisor support.

File: pkg/virt-launcher/virtwrap/converter/converter_test.go (hunk 2 above)
Comment (xiaobo55x) [reply pkg/virt-launcher/virtwrap/converter/converter_test.go:3268]: The name 'qemu' is a little confusing here, it's from libvirt API. Actually, 'qemu' driver mean using 'split mode' IOAPIC when creating a VM. I think Cloud hypervisor should have a similar API to configure the guest IOAPIC mode.

File: pkg/virt-operator/resource/generate/components/validations_generated.go (hunk 3 above)
Comment (xiaobo55x) [reply pkg/virt-operator/resource/generate/components/validations_generated.go:19207]: Yes, should APIC.

File: staging/src/kubevirt.io/api/core/v1/schema_swagger_generated.go (hunk 4 above)
Comment (xiaobo55x) [reply staging/src/kubevirt.io/api/core/v1/schema_swagger_generated.go:599]: Yes, should APIC.

**Comment by xiaobo55x** [comment]
> @xiaobo55x do you see other reasons for exposing those values to KubeVirt API? Otherwise, those values could be set when TDX is requested.
> 
Currently, only TDX have this requirement for some security reason! 
> UPDATE: possibly it should follow something similar to [SEV](https://github.com/kubevirt/kubevirt/pull/6166). Hence, when you request TDX, then the ioapic and pic flags can be set but we don't expose every single option to the end-user.

Good idea! From the performance perspective, the default ioapic configuration is the best choice. 
I will update the patch with your suggestion later!

Thanks,
Haibo

File: api/openapi-spec/swagger.json (hunk 1 above)
Comment (optimistyzy) [reply api/openapi-spec/swagger.json:15844]: > Currently, the patch only consider using libvirt to start the VM. More works were needed to cover the Cloud hypervisor support.

Got it. We should consider RUST based VMM later.

File: pkg/virt-operator/resource/generate/components/validations_generated.go [hunk 5]
```diff
@@ -9651,6 +9668,15 @@ var CRDsValidation map[string]string = map[string]string{
                         discovery. Defaults to false
                       type: boolean
                   type: object
+                pic:
+                  description: PIC enables/disables legacy Programmable Interrupt
+                    Controller.
+                  properties:
+                    enabled:
+                      description: Enabled determines if the feature should be enabled
+                        or disabled on the guest. Defaults to true.
```
Comment (optimistyzy) [inline pkg/virt-operator/resource/generate/components/validations_generated.go:9677]: Could we revise this kind of description?  "Enabled determines if..." It seems not easy for me to understand.

File: api/openapi-spec/swagger.json (hunk 1 above)
Comment (optimistyzy) [inline api/openapi-spec/swagger.json:15844]: APIO -> APIC

**Comment by xiaobo55x** [comment]
@optimistyzy Thanks for the review! 

According to the community feedbacks, the ioapic and pic APIs were only dedicated to TDX, so no need to expose to end user any more. There two patch should be dropped and new series will be sent out later!

**Comment by xiaobo55x** [comment]
Should be re-designed! New series will be sent out later.

**Pull request closed without merging**